processing/src/test/java/org/apache/druid/query/ChainedExecutionQueryRunnerTest.java

```diff
@@ -344,6 +358,151 @@ public void testSubmittedTaskType()
     Assert.assertEquals(runners, actual);
   }
 
+  @Test(timeout = 10_000L)
+  public void testPerSegmentTimeout()
+  {
+    QueryRunner<Integer> slowRunner = (queryPlus, responseContext) -> {
+      try {
+        Thread.sleep(500);
+        return Sequences.of(2);
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    };
+    QueryRunner<Integer> fastRunner = (queryPlus, responseContext) -> Sequences.of(1);
+
+    QueryWatcher watcher = EasyMock.createStrictMock(QueryWatcher.class);
+    watcher.registerQueryFuture(
+        EasyMock.anyObject(),
+        EasyMock.anyObject()
+    );
+    EasyMock.expectLastCall().anyTimes();
+    EasyMock.replay(watcher);
+
+    ChainedExecutionQueryRunner chainedRunner = new ChainedExecutionQueryRunner<>(
+        processingPool,
+        watcher,
+        Arrays.asList(slowRunner, fastRunner)
+    );
+    TimeseriesQuery query = Druids.newTimeseriesQueryBuilder()
+                                  .dataSource("test")
+                                  .intervals("2014/2015")
+                                  .aggregators(Collections.singletonList(new CountAggregatorFactory("count")))
+                                  .context(
+                                      ImmutableMap.of(
+                                          QueryContexts.PER_SEGMENT_TIMEOUT_KEY, 100L,
+                                          QueryContexts.TIMEOUT_KEY, 5_000L
+                                      )
+                                  )
+                                  .build();
+    Sequence seq = chainedRunner.run(QueryPlus.wrap(query));
+
+    List<Integer> results = null;
+    Exception thrown = null;
+    try {
+      results = seq.toList();
+    }
+    catch (Exception e) {
+      thrown = e;
+    }
+
+    Assert.assertNull("No results expected due to timeout", results);
+    Assert.assertNotNull("Exception should be thrown", thrown);
+    Assert.assertTrue(
+        "Should be QueryTimeoutException or caused by it",
+        Throwables.getRootCause(thrown) instanceof QueryTimeoutException
```

Comment:
Since `QueryTimeoutException` can be thrown from multiple places depending on what timeout kicks in, could you also assert on the error message so we know which code path triggered it?